app/controllers/jobs_controller.rb

```diff
@@ -5,40 +5,80 @@ def index
   end
 
   def new
-    @company = Company.find(params[:company_id])
-    @job = Job.new()
+    @categories = Category.all
+    @company    = Company.find(params[:company_id])
+    @job        = Job.new()
   end
 
   def create
-    @company = Company.find(params[:company_id])
-    @job = @company.jobs.new(job_params)
+    @categories = Category.all
+    @company    = Company.find(params[:company_id])
+    @job        = @company.jobs.new(job_params)
     if @job.save
       flash[:success] = "You created #{@job.title} at #{@company.name}"
       redirect_to company_job_path(@company, @job)
     else
+      @error = "A job must have a Title, level of interest, city, and category!"
       render :new
     end
   end
 
   def show
-    @job = Job.find(params[:id])
+    @job      = Job.find(params[:id])
+    @company  = Company.find(params[:company_id])
+    @category = category
+    @comments = @job.comments.all
+    @comment  = @job.comments.new
+    @comment.job_id = @job.id
   end
 
   def edit
-    # implement on your own!
+    @company = Company.find(params[:company_id])
+    @job = @company.jobs.find(params[:id])
+    @categories = Category.all
   end
 
   def update
-    # implement on your own!
+    @company = Company.find(params[:company_id])
+    @job = @company.jobs.find(params[:id])
+    @job.update(job_params)
+    flash[:success] = "You updated #{@job.title} at #{@company.name}"
+
+    redirect_to company_job_path(@company, @job)
   end
 
   def destroy
-    # implement on your own!
+    @company = Company.find(params[:company_id])
+    @job = @company.jobs.find(params[:id])
+    @job.destroy
+
+    redirect_to company_jobs_path(@company)
+  end
+
+  def root
+    if params[:sort] == 'location'
+      @city_jobs = Job.location
```

Comment:
i'd suggest a better name for this `location` method. it's not easy to read and quickly know what data this method returns